lib/bolt/module_installer/puppetfile.rb

```diff
@@ -0,0 +1,121 @@
+# frozen_string_literal: true
+
+require 'puppetfile-resolver'
+
+require 'bolt/error'
+require 'bolt/module_installer/puppetfile/forge_module'
+require 'bolt/module_installer/puppetfile/git_module'
+
+# This class manages the logical contents of a Puppetfile. It includes methods
+# for parsing and generating a Puppetfile.
+#
+module Bolt
+  class ModuleInstaller
+    class Puppetfile
+      attr_reader :modules
+
+      def initialize(modules = [])
+        @modules = modules
+      end
+
+      # Loads a Puppetfile and parses its modules.
+      #
+      def self.parse(path, skip_unsupported_modules: false)
+        return new unless path.exist?
+
+        begin
+          parsed = ::PuppetfileResolver::Puppetfile::Parser::R10KEval.parse(File.read(path))
+        rescue StandardError => e
+          raise Bolt::Error.new(
+            "Unable to parse Puppetfile #{path}: #{e.message}",
+            'bolt/puppetfile-parsing'
+          )
+        end
+
+        unless parsed.valid?
+          raise Bolt::ValidationError, <<~MSG
+            Unable to parse Puppetfile #{path}:
+            #{parsed.validation_errors.join("\n\n")}.
+            This may not be a Puppetfile managed by Bolt.
+          MSG
+        end
+
+        new(to_modules(parsed.modules, path, skip_unsupported_modules))
+      end
+
+      # Converts PuppetfileResolver::Document::*Module to Bolt module objects
+      #
+      private_class_method def self.to_modules(modules, path, skip_unsupported)
```

Comment:
Can this just get inlined to the function above since that's the only place it's used? This seems so particular to dealing with modules from the parser that I don't think it makes sense outside of the `parsed` function.